src/ucp/core/ucp_worker.c

```diff
@@ -357,18 +358,44 @@ static void ucp_worker_init_device_atomics(ucp_worker_h worker)
     }
 }
 
+static void ucp_worker_init_guess_atomics(ucp_worker_h worker)
+{
+    ucp_context_h context = worker->context;
+    ucp_rsc_index_t rsc_index;
+    uint64_t accumulated_flags = 0;
+
+    for (rsc_index = 0; rsc_index < context->num_tls; ++rsc_index) {
+	accumulated_flags |= worker->iface_attrs[rsc_index].cap.flags;
+    }
+
+    if ((accumulated_flags & UCT_IFACE_FLAG_ATOMIC_CPU) &&
```

Comment:
i think checking for CPU flag is not really needed. it would be enough to:
if (accum_flags & ATOMIC_DEVICE) { init_device() } else { init_cpu() }